navigation_and_routing/lib/src/screens/scaffold_body.dart

```diff
@@ -0,0 +1,54 @@
+// Copyright 2021, the Flutter project authors. Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+import 'package:flutter/material.dart';
+
+import '../routing/parsed_route.dart';
+import '../screens/settings.dart';
+import '../util/fade_transition_page.dart';
+import 'authors.dart';
+import 'books.dart';
+
+/// Displays the contents of the body of [BookstoreScaffold]
+class BookstoreScaffoldBody extends StatelessWidget {
+  static GlobalKey<NavigatorState> navigatorKey = GlobalKey<NavigatorState>();
+  final ParsedRoute currentRoute;
+
+  const BookstoreScaffoldBody({
+    required this.currentRoute,
+    Key? key,
+  }) : super(key: key);
+
+  @override
+  Widget build(BuildContext context) {
+    return Navigator(
+      key: navigatorKey,
+      onPopPage: (route, dynamic result) => route.didPop(result),
+      pages: [
+        if (currentRoute.pathTemplate.startsWith('/authors'))
+          const FadeTransitionPage<void>(
+            key: ValueKey('authors'),
+            child: AuthorsScreen(),
+          )
+        else if (currentRoute.pathTemplate.startsWith('/settings'))
+          const FadeTransitionPage<void>(
+            key: ValueKey('settings'),
+            child: SettingsScreen(),
+          )
+        else if (currentRoute.pathTemplate.startsWith('/books') ||
+            currentRoute.pathTemplate == '/')
+          FadeTransitionPage<void>(
+            key: const ValueKey('books'),
+            child: BooksScreen(currentRoute: currentRoute),
+          )
+        //  TODO: determine why the Navigator is built with empty pages when the user is signed out...
```

Comment:
We should figure this out before merging this pr

Reply:
I don't think the framework is doing what I expect here. I'm replacing one page with another, but the `child` for both pages is building during the same build.

Calling `signOut()` changes RouteState from `/settings` to `/signin`, which replaces the Page that builds the BookstoreScaffoldBody (whose parent widget is BookstoreScaffold) with the page that builds the SignInScreen. But I can see the `build()` methods for both of these widgets happening in the same build when I set a breakpoint.

I don't expect the `build()` method for `BookstoreScaffoldBody` to happen after `signOut()`, since it's Page was removed from the the top-level Navigator's pages list, if that makes sense.

Reply:
It is probably due to page transition, before the page transition ends, both pages will be built because they are on the screen at the same time

Reply:
It looks like using InheritedNotifier for RouteStateScope causes the build method for BookstoreScaffoldBody to happen when the route changes, so it will handle whenever the global route changes.  I think this is a scenario where more targeted nested routing is needed. Commit 4138241 adds a comment to explain why this is necessary here.